### PR TITLE
[Feature] batchwise histogram 2d

### PR DESCRIPTION
**Pull request by connorlee77**
Add histogram and histogram2d (joint histogram) into kornia. Marginal and joint pdfs calculated using [kernel density estimation](https://en.wikipedia.org/wiki/Multivariate_kernel_density_estimation) to keep memory and time complexity costs down.

### CODE REVIEW

**Comment by connorlee77** [comment]
https://github.com/kornia/kornia/issues/413#issuecomment-605432944

**Comment by connorlee77** [comment]
@edgarriba How should I fix these tests? Doesn't seem like its any of my added code thats failing.

File: kornia/feature/histogram.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+def marginalPdf(values: torch.Tensor, bins: torch.Tensor, sigma: float = 0.7) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Args:
+        values: (torch.Tensor), shape [BxN]
+        bins: (torch.Tensor), shape [NUM_BINS]
+        sigma: (torch.Tensor), scalar, gaussian smoothing factor
+    Returns:
+        pdf: (torch.Tensor), shape [BxN]
+        kernel_values: (torch.Tensor), shape [BxNxNUM_BINS]"""
+
```
Comment (ducha-aiki) [inline kornia/feature/histogram.py:25]: Add shape checks, please. E.g. https://github.com/kornia/kornia/blob/master/kornia/geometry/linalg.py#L48

File: kornia/feature/histogram.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    residuals = values - bins.unsqueeze(0).unsqueeze(0)
+    kernel_values = torch.exp(-0.5 * (residuals / sigma).pow(2))
+
+    pdf = torch.mean(kernel_values, dim=1)
+    normalization = torch.sum(pdf, dim=1).unsqueeze(1) + EPSILON
```
Comment (ducha-aiki) [inline kornia/feature/histogram.py:30]: epsilon is better to be input with default value

File: kornia/feature/histogram.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    return (pdf, kernel_values)
+
+
+def jointPdf(kernel_values1: torch.Tensor, kernel_values2: torch.Tensor) -> torch.Tensor:
+    """Args:
+        kernel_values1: (torch.Tensor), shape [BxNxNUM_BINS]
+        kernel_values2: (torch.Tensor), shape [BxNxNUM_BINS]
+    Returns:
+        pdf: (torch.Tensor), shape [BxNUM_BINSxNUM_BINS]"""
+
```
Comment (ducha-aiki) [inline kornia/feature/histogram.py:42]: Shape checks

File: kornia/feature/histogram.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    density estimation which requires a bandwidth (smoothing) parameter.
+
+    Args:
+        x: (torch.Tensor), shape [BxN]
+        bins: (torch.Tensor), shape [NUM_BINS]
+        bandwidth: (torch.Tensor), scalar, gaussian smoothing factor
+    Returns:
+        pdf: (torch.Tensor), shape [BxNUM_BINS]"""
+
+    x = x * 255
```
Comment (ducha-aiki) [inline kornia/feature/histogram.py:62]: Why 255? It should be an input parameter

File: kornia/feature/histogram.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    Args:
+        x1: (torch.Tensor), shape [BxN1]
+        x2: (torch.Tensor), shape [BxN2]
+        bins: (torch.Tensor), shape [NUM_BINS]
+        bandwidth: (torch.Tensor), scalar, gaussian smoothing factor
+    Returns:
+        pdf: (torch.Tensor), shape [BxNUM_BINSxNUM_BINS]"""
+
+    x1 = x1 * 255
```
Comment (ducha-aiki) [inline kornia/feature/histogram.py:81]: Same for 255

File: test/feature/test_histogram.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        pdf = histogram(inp, bins, 1)
+        assert pdf.shape == (8, 128)
+
+    def test_gradcheck(self, device):
+        inp = torch.ones(8, 32, device=device)
+        inp = utils.tensor_to_gradcheck_var(inp)  # to var
+        bins = torch.linspace(0, 255, 128).to(device)
+
+        assert gradcheck(histogram, (inp, bins, 1), raise_exception=True)
+
```
Comment (ducha-aiki) [inline test/feature/test_histogram.py:31]: Add some real test please, besides sanity checks. With some toy input data and expected value.

File: kornia/feature/histogram.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import os
+import numpy as np
+
+import torch
+import torch.nn as nn
+
+import matplotlib.pyplot as plt
+
+from PIL import Image
+from torchvision import transforms
```
Comment (edgarriba) [inline kornia/feature/histogram.py:12]: remove all non torch dependencies

File: kornia/feature/histogram.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+EPSILON = 1e-10
```
Comment (edgarriba) [inline kornia/feature/histogram.py:14]: I would add this into the function signature

File: kornia/feature/histogram.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def marginalPdf(values: torch.Tensor, bins: torch.Tensor, sigma: float = 0.7) -> Tuple[torch.Tensor, torch.Tensor]:
```
Comment (edgarriba) [inline kornia/feature/histogram.py:17]: no camelCase, use snake_case: `margina_pdf` plus add some minimal description about what the actual function does

File: kornia/feature/histogram.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    kernel_values = torch.exp(-0.5 * (residuals / sigma).pow(2))
+
+    pdf = torch.mean(kernel_values, dim=1)
+    normalization = torch.sum(pdf, dim=1).unsqueeze(1) + EPSILON
+    pdf = pdf / normalization
+
+    return (pdf, kernel_values)
+
+
+def jointPdf(kernel_values1: torch.Tensor, kernel_values2: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/feature/histogram.py:36]: snake_case

File: kornia/feature/histogram.py (hunk 9 above)
Comment (edgarriba) [reply kornia/feature/histogram.py:17]: I guess sigma could be also a tensor that can be optimized, right ?

File: kornia/feature/histogram.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        pdf: (torch.Tensor), shape [BxNUM_BINSxNUM_BINS]"""
+
+    joint_kernel_values = torch.matmul(kernel_values1.transpose(1, 2), kernel_values2)
+    normalization = torch.sum(joint_kernel_values, dim=(1, 2)).view(-1, 1, 1) + EPSILON
+    pdf = joint_kernel_values / normalization
+
+    return pdf
+
+
+def histogram(x: torch.Tensor, bins: torch.Tensor, bandwidth: float = 0.7) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/feature/histogram.py:50]: bandwidth can be tensor too

File: kornia/feature/histogram.py (hunk 4 above)
Comment (edgarriba) [reply kornia/feature/histogram.py:62]: agree, for convention in kornia we assume [0-1] floating point input ranges

**Comment by edgarriba** [comment]
@connorlee77 we are releasing v0.2.1 next Tuesday. Would be great to include this feature. Do you think you can make it for the weekend ?

**Comment by connorlee77** [comment]
@edgarriba yes, I'll get on it

**Comment by connorlee77** [comment]
@edgarriba @ducha-aiki addressed the issues for this PR

**Comment by edgarriba** [comment]
@connorlee77 the code looks. Just realized now that you propose to add it in `kornia.feature`. Wouldn't be more appropriate in `kornia.color` ?

**Comment by connorlee77** [comment]
> @connorlee77 the code looks. Just realized now that you propose to add it in `kornia.feature`. Wouldn't be more appropriate in `kornia.color` ?

Hmm perhaps. Although it looks to me like most functionalities in kornia.color are preprocessing/color conversions. 

I can change it in a few hours. When's the deadline for this PR to make it into the new kornia update?

**Comment by edgarriba** [comment]
@connorlee77 we still have a couple of days for the release.

after quick research,
opencv have it in imgproc->histrograms: https://docs.opencv.org/4.3.0/d6/dc7/group__imgproc__hist.html
scikit in `exposure`: https://scikit-image.org/docs/dev/api/skimage.exposure.html#module-skimage.exposure

scikit have also all the adjust* group operators in there. Maybe it's worth creating similar module, my guess is that all the normalisation functions could be moved there too. /cc @ducha-aiki 

@justusschock what do you think ? you have some normalization stuff that have to be ported probably in there too.

**Comment by edgarriba** [comment]
@ducha-aiki cool ! @connorlee77 can you make those changes ?

**Comment by connorlee77** [comment]
@edgarriba moved it to kornia.color. some merge conflicts popped up after autopeping that I fixed. seems like random.py still has a conflict according to this. However, it seems good on my end.

**Comment by edgarriba** [comment]
@did you rebase against master ?

**Comment by connorlee77** [comment]
> @did you rebase against master ?

Yeah I did

**Comment by edgarriba** [comment]
@connorlee77 weird, because I see a lot of unrelated commits

**Comment by edgarriba** [comment]
@connorlee77 been checking in local and I couldn't figure out the problem with your commits historial. My guess is that rebase hasn't succeed at some points making nonlinear commits history. Since you have pretty much isolated changed (basically adding files), I would suggest to create the PR from scratch with a fresh rebase from master and add in a single commit your functionalities. Otherwise this could complicate the whole commits history for others.
